Wizard refactoring

levaja opened the pull request:
This PR address issue #147 . User registration wizard is refactored in order to enable faster user+project registration.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/handlers/StartupUIThread.java:
```diff
@@ -20,29 +20,30 @@
 	/** The preferences. */
 	private Preferences preferences;
 
+	/** Whether user has cancelled first registration. */
```
What does first registration mean?

levaja replied:
It's user+project registration, I'll fix this to make it more precise.

Inventitech commented:
Well done! The new pages look awesome and it's so much faster to register.

I found these issues:
The last pages still don't seem to use the same code. (In the successful scenario, the title is bold and in the error case it's not).

Even after a successful registration, the flag "monitor this workspace" is still not force-set, which can lead to WD being inactive inspite of just registering successfully.

A few nitpicks:
Typo: "Welcome! This registrations"
-> "Welcome! Registration is fun and takes just 3 minutes"

link "amazing prizes", and change "a detailed report" to "a detailed report on your development behaviour" and link http://www.testroots.org/reports/sample_watchdog_report.pdf

Selecting "No, I have a User-ID does not change the number of pages"

Change "All projects" -> "Do all Eclipse projects in this workspace belong to one 'larger', project"?

levaja commented:
>The last pages still don't seem to use the same code. (In the successful scenario, the title is bold and in the error case it's not).

They are using the same methods from UIUtils (createSuccessMessage and createFailureMessage), but they have some small printing on their own, due to the difference in the displaying messages they have. However, I have an idea how to make it even more better, so I would fix that as well.

>Even after a successful registration, the flag "monitor this workspace" is still not force-set, which can lead to WD being inactive inspite of just registering successfully.

Thanks for pointing this out, I've missed that. I will look into this now.

>Typo: "Welcome! This registrations"
-> "Welcome! Registration is fun and takes just 3 minutes"

Fixed.

>Selecting "No, I have a User-ID does not change the number of pages"

It doesn't need to change - user registration now has 1 page (after welcome) and providing user id has just one additional (confirmation) page, so the number of pages in wizard stays the same. It's not the case for project registration, so that number changes.

>Change "All projects" -> "Do all Eclipse projects in this workspace belong to one 'larger', project"?

Fixed.

levaja commented:
Should be fine now.

Inventitech commented:
Hi Igor,

good work again. I'm sure many of the bugs that I find here are old bugs and not regressions caused by your code.

If a server communication problem happened, it is not possible to leave the wizard on the last page. If the registration was not successful, cancel should be active, and no changes made to the preferences.

![image](https://cloud.githubusercontent.com/assets/334256/5920711/e321f602-a63e-11e4-8711-6b38c0a35d69.png)

levaja commented:
Hi Moritz,

Thanks for all the feedback! Fixed this one as well.

levaja commented:
>Rather strange for me, even after successful registration the WD icon and status text still tell me that WD is not running, but the option to "monitor the workspace" is ticked.

Can you tell me the scenario in which this happens? Doesn't happen to me.

Inventitech commented:
OK, I just tested the functionality and it seems to me that everything we specified is working. I am now going to take a look at the code.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/handlers/StartupUIThread.java:
```diff
@@ -20,29 +20,30 @@
 
-		if (!WatchDogUtils.isEmpty(preferences.getUserid())) {
-			// In case the user aborted the preference dialog with cancel,
+		if (!WatchDogUtils.isEmpty(preferences.getUserid())
```
I don't understand this condition anymore. I assume the comment is not up-to-date anymore? Why are you checking and modifying userRegistrationCancelled?

levaja replied:
userRegistrationCancelled is a boolean flag that I've added. It ensures that if the user cancelled wizard after registering user-id, but before registering project, he would not be asked again for project registration (if condition false, ignoring the block).Otherwise, if he changes workspace, he will be offered just to do workspace registration (if condition true, going into block). Is my explanation clear enough?

Inventitech replied:
No.

levaja replied:
Okay, I'll try like this: 
1) First WD installation, user starts user+project registration
2) Finishes user registration part, cancel during project registration (he doesn't want to register current workspace)
3) Flag is set on true, to avoid WD asking user whether or not he would like WD to monitor this workspace (since he cancelled, he obviously doesn't want this).
4) If a user later changes his workspace, this flag would be false, so another workspace will be checked, with offering just project registration wizard.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+
+import org.eclipse.jface.wizard.Wizard;
+
+/** Parent class for User and Project registration wizards. */
```
Base class for ...

levaja replied:
Fixed.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+import org.eclipse.jface.wizard.Wizard;
+
+/** Parent class for User and Project registration wizards. */
+public abstract class RegistrationWizard extends Wizard {
```
call this RegistrationWizardBase ?

levaja replied:
WelcomePage, IdEnteredEndingPage, RegistrationEndingPage are all base classes for the pages in specific wizards, but they don't have Base in name. Should only this be renamed, leave it like it is, or rename everything?

Inventitech replied:
Please rename them, too.

levaja replied:
Renamed everything.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+	protected String projectId;
+
+	/** Maximum number of pages in the current wizard. */
+	protected int totalPageNumber;
```
rename to totalPages (it's an int, so it's clear it's a number)

levaja replied:
Done.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+
+	@Override
+	public boolean canFinish() {
+		FinishableWizardPage currentPage = (FinishableWizardPage) getContainer()
```
What if the current page is not a finishableWizardPage?

levaja replied:
Not possible, FinishableWizardPage is base class for all the pages. Also, this is a piece of code from old versions, just moved from subclasses to this base class.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+	protected ProjectCreatedEndingPage projectedCreatedPage;
+
+	@Override
+	public boolean performFinish() {
```
Fetch preferences once.

levaja replied:
Done.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+	protected int totalPageNumber;
+
+	/** Page with Yes/No question about registering new project ID. */
+	public ProjectWelcomePage projectWelcomePage;
```
Why public?

Inventitech replied:
I am currently also not sure why the wizard base should no about the specific projectWelcomePage. Isn't this a logical flaw?

levaja replied:
When existing project-ID is entered, on the next page (which is also the last page of the wizard, so Finish is enabled, but Back is disabled) it is necessary to fetch entered ID in order to register it with the server. This is only possible with getWizard.projectWelcomePage.getId()

levaja replied:
My previous comment refers to your first comment. For the second, it is arguable - at the moment, these pages are part of both wizards, so it is justified to leave them in a base class. On the other hand, moving them downwards (to subclasses) would be okay if we have some more registrations, but I am guessing it will not happen in the near future. 
My reason for putting these pages there is to be able to access them from (RegistrationWizard) getWizard()

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+		return currentPage.canFinish();
+	}
+
+	/** Getting project-ID. */
```
Please use "Returns project-ID" to make consistent with the exisiting code base.

levaja replied:
Done.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+		return projectId;
+	}
+
+	/** Setting up projectId field. */
```
Please use "Sets project-ID" to make consistent with the exisiting code base.

levaja replied:
Done.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+		this.projectId = projectId;
+	}
+
+	/** Getting total number of pages in current wizard. */
```
Consolidate comment.

levaja replied:
Done.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/RegistrationWizard.java:
```diff
@@ -0,0 +1,76 @@
+		return totalPageNumber;
+	}
+
+	/** (Re)Setting up total number of pages in current wizard. */
```
Consolidate comment.

levaja replied:
Done.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/WelcomePage.java:
```diff
@@ -95,43 +97,47 @@ private Composite createQuestionComposite(final Composite parent) {
-				setTitle(title + " (1/1)");
+				dynamicContent = createWelcomeComposite(parent);
+				setTitle(title
+						+ " ("
```
By pulling this calculation of the page number up into the base class, you avoid a lot cloned and difficult to read code.

levaja replied:
This (and equivalent comments in other classes) I don't quite understand. WelcomePage is base class for both UserWelcomePage and ProjectWelcomePage. Field currentPageNumber is located in the base class of all pages (FinishableWizardPage), but title can change during registration process on some pages, so that's why each page is responsible for their title. Can you elaborate a bit more what exactly do you have in mind?

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/projectregistration/ProjectIdEnteredEndingPage.java:
```diff
@@ -2,31 +2,32 @@
  * the server is not reachable, the user can exit here.
  */
-class ProjectIdEnteredEndingPage extends IdEnteredEndingPage {
+public class ProjectIdEnteredEndingPage extends IdEnteredEndingPage {
```
class is called ProjectXX but actually works for all RegistrationWizards?

levaja replied:
True. Technically, in both registration wizards, there is a project registration. Therefore, maybe "UserRegistrationWizard" should be renamed to something like UserProjectRegistrationWizard.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/projectregistration/ProjectRegistrationPage.java:
```diff
@@ -139,9 +142,15 @@ public void validateFormInputs() {
 		}
 
-		setTitle(TITLE + " (2/3)");
+		setTitle(TITLE + " (" + currentPageNumber + "/"
```
By pulling this calculation of the page number up into the base class, you avoid a lot cloned and difficult to read code.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/projectregistration/ProjectRegistrationPage.java:
```diff
@@ -139,9 +142,15 @@ public void validateFormInputs() {
+				+ ((RegistrationWizard) getWizard()).getTotalPageNumber() + ")");
 		if (shouldSkipProjectSliderPage()) {
-			setTitle(TITLE + " (2/2)");
+			setTitle(TITLE
```
By pulling this calculation of the page number up into the base class, you avoid a lot cloned and difficult to read code.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/WelcomePage.java:
```diff
@@ -95,43 +97,47 @@ private Composite createQuestionComposite(final Composite parent) {
-				dynamicContent = createWelcomeComposite(parent);
-				setTitle(title + " (1/3)");
+				dynamicContent = createLoginComposite(parent);
+				int total = currentRegistration.equals("User") ? ((RegistrationWizard) getWizard())
```
Such behaviour should never depend on a string comparison. What you really want is two separate classes. The second page number calculation looks like the total number of pages is somehow wrong.

levaja replied:
I agree that this doesn't look nice, but this was the only solution if I wanted to avoid making project registration classes for user registration wizard. Furthermore, this is the base class (radio button is created in this class for both welcome pages), so you maybe think about moving this to child classes (UserWelcomePage and ProjectWelcomePage) ?

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/projectregistration/ProjectCreatedEndingPage.java:
```diff
@@ -58,11 +79,68 @@ protected void makeRegistration() {
 		successfulRegistration = true;
-		((ProjectRegistrationWizard) getWizard()).projectId = id;
+
+		((RegistrationWizard) getWizard()).setProjectId(id);
```
Looks unsafe.

levaja replied:
Actually, it was unsafe before, with accessing field in project wizard. Now with the base class, setProjectId(id) method from this class sets the field in the same class. Cast operation is safe, since the wizard can only be RegistrationWizard.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/projectregistration/ProjectWelcomePage.java:
```diff
@@ -2,26 +2,34 @@
+	public IWizardPage getPreviousPage() {
+		return getWizard().getPreviousPage(this);
+	}
 }
```
file does not end with \n.

levaja replied:
Fixed.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/projectregistration/ProjectCreatedEndingPage.java:
```diff
@@ -58,11 +79,69 @@ protected void makeRegistration() {
+
+	private void createPageContent() {
+		setTitle(windowTitle);
+		if (isThisProjectWizzard()) {
```
Can't you make the page wizard-agnostic by providing subclasses that are added to the correct wizards?

levaja replied:
Well, that was one thing I have been trying to avoid - to make more (duplicate or redundant) classes than needed. In the current scenario, there is no mixing up classes between the packages, but project registration classes specific for user registration wizard break the logic a bit imo. Therefore, in this case, I've opted for using this simple method.

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/userregistration/UserRegistrationPage.java:
```diff
@@ -78,7 +83,7 @@ private Composite createRegistrationComposite(Composite parent) {
 		UIUtils.createLabel("Your Programming Experience: ", composite);
 		experienceDropDown = UIUtils.createComboList(composite, formValidator,
-				new String[] { "< 1 year", "1-2 years", "3-6 years",
+				new String[] { "", "< 1 year", "1-2 years", "3-6 years",
```
nice :)

levaja replied:
Thanks :)

Inventitech commented on WatchDogEclipsePlugin/WatchDog/src/nl/tudelft/watchdog/ui/wizards/userregistration/UserWelcomePage.java:
```diff
@@ -1,27 +1,87 @@
+			welcomeTextLabel.getParent().layout();
+			welcomeTextLabel.getParent().update();
+
+			String descriptionText = "It measures how you write Java code and tests. We never do anything bad with <a href=\"http://www.testroots.org/testroots_watchdog.html#details\">your purely numerical data</a>.\nYou can win <a href=\"http://www.testroots.org/testroots_watchdog.html#prizes\">amazing prizes</a>, <a href=\"http://www.testroots.org/reports/sample_watchdog_report.pdf\">a detailed report o ...
```
only link detailed report, please :)

levaja replied:
Fixed.
